File: python/app/core/calculator.py
from dataclasses import dataclass

from app.core.models import SymbolDilutedCost, Transaction, TransactionType


@dataclass
class _Accumulator:
    shares: float = 0.0
    net_diluted_cost: float = 0.0


def calculate_diluted_costs(transactions: list[Transaction]) -> list[SymbolDilutedCost]:
    symbols: dict[str, _Accumulator] = {}

    for tx in transactions:
        if tx.symbol not in symbols:
            symbols[tx.symbol] = _Accumulator()

        acc = symbols[tx.symbol]

        if tx.type == TransactionType.Trades:
            acc.shares += tx.quantity

        acc.net_diluted_cost += tx.net_amount

    results: list[SymbolDilutedCost] = []
    for symbol in sorted(symbols):
        acc = symbols[symbol]
        diluted_cost_per_share = acc.net_diluted_cost / acc.shares if acc.shares != 0 else None
        results.append(SymbolDilutedCost(
            symbol=symbol,
            shares=acc.shares,
            net_diluted_cost=round(acc.net_diluted_cost, 2),
            diluted_cost_per_share=round(diluted_cost_per_share, 2) if diluted_cost_per_share is not None else None,
        ))

    return results

File: src/lib/portfolio/Calculator.ts
import {type PortfolioRow, type Transaction, TransactionType} from "./Portfolio";

interface Accumulator {
  shares: number;
  netDilutedCost: number;
}

export function calculateDilutedCosts(transactions: Transaction[]): PortfolioRow[] {
  const symbols = new Map<string, Accumulator>();

  for (const tx of transactions) {
    let acc = symbols.get(tx.symbol);
    if (!acc) {
      acc = {shares: 0, netDilutedCost: 0};
      symbols.set(tx.symbol, acc);
    }

    if (tx.type === TransactionType.Trades) {
      acc.shares += tx.quantity!;
    }

    acc.netDilutedCost += tx.netAmount;
  }

  const results: PortfolioRow[] = [];
  for (const symbol of [...symbols.keys()].sort()) {
    const acc = symbols.get(symbol)!;
    const dilutedCostPerShare = acc.shares !== 0 ? acc.netDilutedCost / acc.shares : null;
    results.push({
      symbol: symbol,
      shares: acc.shares,
      netDilutedCost: acc.netDilutedCost,
      dilutedCostPerShare: dilutedCostPerShare,
    });
  }

  return results;
}
